src/cargo/core/compiler/mod.rs

```diff
@@ -1094,23 +1094,35 @@ fn build_deps_args<'a, 'cfg>(
             }
             let mut v = OsString::new();
 
-            // Unfortunately right now Cargo doesn't have a great way to get a
-            // 1:1 mapping of entries in `dependencies()` to the actual crate
-            // we're depending on. Instead we're left to do some guesswork here
-            // to figure out what `Dependency` the `dep` unit corresponds to in
-            // `current` to see if we're renaming it.
-            //
-            // This I believe mostly works out for now, but we'll likely want
-            // to tighten up this in the future.
-            let name = current
-                .pkg
-                .dependencies()
-                .iter()
-                .filter(|d| d.matches_ignoring_source(dep.pkg.package_id()))
-                .filter_map(|d| d.rename())
-                .next();
-
-            v.push(name.unwrap_or(&dep.target.crate_name()));
+            let deps = {
+                let a = current.pkg.package_id();
+                let b = dep.pkg.package_id();
+                if a == b {
+                    &[]
+                } else {
+                    cx.resolve.dependencies_listed(a, b)
+                }
+            };
+
+            let mut name = None;
+            let crate_name = dep.target.crate_name();
+            for n in deps.iter().map(|d| d.rename()) {
+                match (n, name) {
+                    (None, None) => continue,
+                    (Some(n), Some(prev)) if n == prev => continue,
+                    (None, Some(n)) if n == crate_name => continue,
+                    (Some(n), None) => {
+                        name = Some(n);
+                        continue
+                    }
+                    _ => {}
+                }
+                bail!("multiple dependencies listed for the same crate must \
```

Comment:
I believe this is correct, but the logic here is kinda hard to follow. Perhaps the following would be easier to undestand?

```
let mut names = deps.iter().map(|d| d.rename().unwrap_or(crate_nam));
let name = names.next().unwrap();
for n in names {
    if n != name { bail!("multiple dependencies...") }
}
```